### DISCUSSION TITLE
Simple distance helper functions?

### DISCUSSION THREAD

**Discussion by okaneco** [Ogeon/palette#275] [Ideas]
I find myself rewriting boilerplate often enough that maybe it'd be nice to have upstream. The `ColorDifference` trait exists but I don't always need the expensive CIEDE2000 calculation. Sometimes I'd like a more simple Euclidean distance/norm or the squared distance for comparing to a tolerance value.

This is a sketch of what I had in mind.
```rust
fn distance_squared(&self, other: &Self) -> T {
    (self.red - other.red).powi(2)
        + (self.green - other.green).powi(2)
        + (self.blue - other.blue).powi(2)
}

fn distance(&self, other: &Self) -> T {
    ((self.red - other.red).powi(2)
        + (self.green - other.green).powi(2)
        + (self.blue - other.blue).powi(2))
    .sqrt()
}
```

Perhaps this can be added to the `ColorDifference` trait?

There may be other "general enough" functions that people find themselves re-implementing similar to these which would carry their weight as additions to the library.

**Comment by Ogeon** [comment]
That's a nice idea! I have been thinking that I would like to replace `ColorDifference` with one or more traits that are more specific about their definitions. I may be wrong, but CIEDE2000 is specific for `Lab` and would require an implicit conversion for anything else. So I would like to maybe move that to `Lab` itself to make that clear. I don't have a clear idea yet, but this post highlights another reason why I would like something more well defined. So there could be something like `EuclideanDistance` for these `distance` and `distance_squared` and maybe rename or replace the current `ColorDifference`.

**Reply by Ogeon** [reply]
I feel like I need to explain a bit more. I'm generally trying to avoid being opinionated, since that has proven to be a mistake. Multiple times, actually. There's always some application specific reason to do things another way, so I always end up loosening the requirements and adding more options. I feel like the concept "color difference" is not specific enough for me to pick a single canonical definition. I would rather give number of options and not necessarily favor either of them over the other.

I should have caught this earlier, when the trait was first added. Sorry about that. I didn't really have the same perspective of it then, so that may have been part of the reason why I didn't think of it. It's a mindset that has crystalized more during the past half year or so, even though it has been the direction since the second release ever of the library.

So, good idea, but let's give them their own trait. 👍

**Reply by okaneco** [reply]
No worries, I agree with all of that. With hindsight and experience the past couple of years, current me would not have done things quite the same way past me did. :smile:

`EuclideanDistance` sounds good. I was trying to think of any other low hanging fruit that would fit but I think this is it for now. Almost all the math functions I need are already well accounted for.

**Reply by Ogeon** [reply]
Good to hear I haven't lost it completely yet! 😁 I agree that these two functions sound good for this trait. I like them small and specific.
